src/bin/cql-stress-cassandra-stress/settings/param/types.rs

```diff
@@ -298,9 +299,86 @@ impl Parsable for ConnectionsPerShard {
     }
 }
 
+/// A ratio map which should match the following pattern:
+/// (<item1>=<f64>,<item2>=<f64>,...,<item_n>=<f64>)
+///
+/// Requirements:
+/// - Items have to be unique -> "(foo=1,foo=2)" parsing should fail
+/// - User needs to specify at least one item's ratio -> "()" parsing should fail
+/// - Weights cannot be negative -> "(foo=-1)" parsing should fail
+/// - Weights cannot sum up to 0 -> "(foo=0,bar=0)" parsing should fail
+///
+/// Last 3 requirements are introduced so creating a [rand_distr::WeightedIndex] with
+/// [rand_distr::WeightedIndex::new] from iterator of f64 values does not fail.
+pub struct RatioMap;
+
+impl RatioMap {
+    fn parse_item_weight(s: &str) -> Result<(&str, f64)> {
+        let (item, weight) = {
+            let mut iter = s.split('=').fuse();
+            match (iter.next(), iter.next(), iter.next()) {
+                (Some(cmd), Some(w), None) => (cmd, w),
+                _ => anyhow::bail!("Item weight specification should match pattern <item>=<f64>"),
+            }
+        };
+
+        let weight = weight.parse::<f64>()?;
+        anyhow::ensure!(weight >= 0f64, "Item weight cannot be negative: {}", weight);
+
+        Ok((item, weight))
+    }
+
+    fn do_parse(s: &str) -> Result<HashMap<String, f64>> {
+        // Remove wrapping parenthesis.
+        let arg = {
+            let mut chars = s.chars();
+            anyhow::ensure!(
+                chars.next() == Some('(') && chars.next_back() == Some(')'),
+                "Invalid ratio specification: {}",
+                s
+            );
```

Comment:
Note that if this `ensure!` fails, the resulting error will contain a message with double `Invalid ratio specification: Invalid ratio specification: ...`, because the outer `do_parse` prepends such header too.